src/core/control/tools/ImageHandler.cpp

```diff
@@ -85,6 +83,12 @@ auto ImageHandler::insertImage(GFile* file, double x, double y) -> bool {
     img->setWidth(width * zoom);
     img->setHeight(height * zoom);
 
+    return std::make_tuple(true, img, width * zoom, height * zoom);
+}
```

Comment:
Ok I see we need to talk/think more about this. You're right doing the automatic scaling right when creating the image is wrong as if the user only sets `width` and `aspectratio=false`, `height` will have the wrong value.

Adding the automatic scaling to the second split function `addImageToDocument` is also wrong because then the image will/might get scaled after we did our manual scaling.

So I think we'd need an additional function which does the automatic scaling. Do you agree on this?

With that it would be logical to simply not use automatic scaling in the api function so that scaling is up to the api-user. But the question arises if this doesn't lead to many troubles where a user inserts an image without any scaling and the image is way too big.
A solution for this issue might be to check if any of `width`, `height`, `scale` was set. If so, disable automatic scaling, if not, enable automatic scaling. Still the user might want to insert the image completely unscaled (neither manually nor automatically) -> add another (bool) parameter for controlling automatic scaling (default to `true`)?

So I see the problem, but I'm not sure how to proceed in this issue.

Reply:
The API user could get the width and image of the picture beforehand e.g. using  the ImageMagick `identify` tool (```identify -format "%wx%h" img.png```) and then decide based on the page size and insertion location. That may be slightly unpractical though.
What about introducing optional `maxWidth` and `maxHeight` parameters? I think no automatic (uncontrollable) scaling should ever be applied through the Lua API.

Reply:
- manually get the width and height: I too think that's pretty impractical (especially for plugins which should be portable we'd need to always check if `identify` is available, and lots of `popen` calls are not nice anyhow)
- setting maximums sounds like a nice idea. So it's gonna be something like this
```
autoWidth, autoHeight = calc automatic scaling
ar = keep_aspect_ratio
if manualWidth set {
  width = min(manualWidth, autoWidth)
  if ar {
    set height based on width
  } else {
    height = autoHeight
  }
}
else if manualHeight set {
  height = min(manualHeight, autoHeight)
  if ar {
    set width based on height
  } else {
    width = autoWidth
  }
}
else if both are set {
  height = min(manualHeight, autoHeight)
  width = min(manualWidth, autoWidth)
}
else {
  width = autoWidth
  height = autoheight
}

```
 
The issue here is that if only one dimension is set and the aspect ratio should be kept the other dimension might be too large.

[Not a trivial task, trying to enforce certain dimensions while giving the user as much freedom as possible]

Reply:
The maxWidth and maxHeight settings should take priority.
I would simply calculate like currently, but in the end if width > maxWidth set width = maxWidth (and scale height proportionally if aspect ratio should be preserved) and then if still height > maxHeight set height = maxHeight (and scale width proportionally if aspect ratio should be preserved).

Reply:
Please check 0da200b9f and 9e1a8488b and tell me what you think?
(not sure if it fits in nicely so I keep this one open)

Reply:
Well, 0da200b9f496f6f7eb481585e26c6e8eb7a745bc does some automatic scaling and completely ignores `maxWidth` and `maxHeight`. That can't be right.

Reply:
The trick is that the `width`/`height` passed into the `automaticScaling` function are the dimensions resulting from the manual scaling. Thus, `automaticScaling` will check if the dimensions are too large, if so it scales them down (keeping the current aspect ratio), if not, the dimensions are not being modified.
(At least that were my thoughts and that is what is supposed to happen)

The `maxWidth`/`maxHeight` parameters passed to the api function are already in the calculated `height`/`width` passed to `automaticScaling` (c.f. Lines 2100 - 2118).

Reply:
In lines 2100-2118 I don't see `maxWidth`/`maxHeight` used. Have you pushed all your changes? Please check!

Reply:
Hm, maybe we're talking about different things. To clarify my intention, I've renamed the variables holding the `maxWidth`/`maxHeight` parameter in the code (see 793c7d5c5).
If this is not what you meant, please clarify.

Reply:
Sorry, I should have read the code more carefully. The current implementation looks good. You should always test the API though to see if it really works as expected.

Reply:
No problem, glad that the current state is fine. Overlooking stuff happens to all of us. So far my tests look good (no images that are too big and no modification of images that fit on the page)